File: code/FenwickTree/FenwickTree.cpp
// Binary indexed tree or fenwick tree
// Space Complexity: O(N) for declaring another array of N=size num_of_elements
// Time Complexity: O(logN) for each operation(update and query as well)
// original array for storing values for later lookup
// Part of Cosmos by OpenGenus Foundation
#include <bits/stdc++.h>
using namespace std;
int num_elements;
void update(int x,int delta,vector<int > &bit)
{
	// x&(-x) gives the last set bit in a number x
	for(;x<=num_elements;x+=x&-x)
		bit[x]+=delta;
}
long long query(int x,vector<int> &bit)
{
	long long sum=0;
	// x&(-x) gives the last set bit in a number x
	for(;x>0;x-=x&-x)
		sum+=bit[x];
	return sum;
}
int main()
{
	int num_queries,left_index,right_index;
	cout<<"Enter number of elements\n";
	cin>>num_elements;
	vector<int> container,bit(num_elements+1,0);
	for(int i=0;i<num_elements;i++)
	{
		cout<<"Enter element\n";
		int temp;
		cin>>temp;
		container.push_back(temp);
		update(i+1,temp,bit);
	}
	cout<<"Enter number of queries\n";
	cin>>num_queries;
	for(int i=0;i<num_queries;i++)
	{
		cout<<"Enter start index(1 indexed)\n";
		cin>>left_index;
		cout<<"Enter end index(1 indexed)\n";
		cin>>right_index;
		if(left_index>right_index)
		{
			cout<<"Invalid range\n";
			continue;
		}
		cout<<"Sum in given range is\n";
		cout<<query(right_index,bit)-query(left_index-1,bit)<<endl;
	}
	return 0;
}

File: cosmos-master/code/data_structures/fenwick_tree/FenwickTree.py
# Binary indexed tree or fenwick tree
# Space Complexity: O(N) for declaring another array of N=size num_of_elements
# Time Complexity: O(logN) for each operation(update and query as well)
# original array for storing values for later lookup
# Part of Cosmos by OpenGenus Foundation
array=[]
# array to store cumulative sum
bit=[]
'''
index i in the bit[] array stores the cumulative sum from the index i to i - (1<<r) + 1 (both inclusive),
 where r represents the last set bit in the index i
'''
class FenwickTree:
# To intialize list of num_of_elements+1 size
	def initialize(self,num_of_elements):
		for i in range(num_of_elements+1):
			array.append(0)
			bit.append(0)

	def update(self,x,delta):
		while x<=num_of_elements:
			bit[x]=bit[x]+delta
			# x&(-x) gives the last set bit in a number x
			x=x+(x&-x)

	def query(self,x):
		range_sum=0
		while x>0:
			range_sum=range_sum+bit[x]
			# x&(-x) gives the last set bit in a number x
			x=x-(x&-x)
		return range_sum

fenwick_tree=FenwickTree()
num_of_elements=int(input("Enter the size of list: "))

fenwick_tree.initialize(num_of_elements)

for i in range(num_of_elements):
	# storing data in orginal list
	element=int(input("Enter the list element: "))
	# updating the BIT array
	fenwick_tree.update(i+1,element)

number_of_queries=int(input("Enter number of queries: "))

for i in range(number_of_queries):
	left_index=int(input("Enter left index (1 indexing): "))
	right_index=int(input("Enter right index (1 indexing): "))
	if right_index < left_index:
		print("Invalid range ")
		continue
	print("Sum in range[%d,%d]: "%(left_index,right_index))
	print(fenwick_tree.query(right_index)-fenwick_tree.query(left_index-1))
